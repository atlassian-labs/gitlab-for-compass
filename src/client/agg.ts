import api from '@forge/api';
import { aggResponseErrorHandler } from './agg-response-error-handler';
import { AggOperation } from '../types';

export const aggQuery = async (request: AggOperation) => {
  const response = await api.asApp().requestGraph(request.query, request.variables);
  const responseBody = await response.json();
  // eslint-disable-next-line no-console
  console.log({
    message: 'AGG request',
    requestName: request.name,
    responseStatus: response.status,
    requestId: responseBody?.extensions?.gateway?.request_id,
  });
  return aggResponseErrorHandler(responseBody);
};
