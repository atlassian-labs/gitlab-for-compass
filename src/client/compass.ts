import graphqlGateway, {
  CompassLinkType,
  CompassCreateEventInput,
  Component,
  ComponentPayload,
  CreateCompassComponentExternalAliasInput,
  DeleteCompassComponentExternalAliasInput,
  DetachCompassComponentDataManagerInput,
  GetComponentByExternalAliasInput,
  UpdateCompassComponentDataManagerMetadataInput,
  UpdateComponentInput,
  SdkError,
} from '@atlassian/forge-graphql';
import { ImportableProject, COMPASS_GATEWAY_MESSAGES, Metric } from '../types';
import { EXTERNAL_SOURCE, IMPORT_LABEL } from '../constants';
import { UNKNOWN_EXTERNAL_ALIAS_ERROR_MESSAGE } from '../models/error-messages';
import { AggClientError, GraphqlGatewayError } from '../models/errors';

const throwIfErrors = function throwIfSdkErrors(method: string, errors: SdkError[]) {
  // Checking if any invalid config errors to report.
  // Certain client errors cannot be validated before calling the compass API, but should be raised as user errors.
  const clientErrors = errors.filter((e) => e?.statusCode === 400);
  if (clientErrors.length > 0) {
    console.warn({ message: 'invalid request error', method, errors: clientErrors.map((e) => e.message) });
    throw new AggClientError(clientErrors.map((e) => e.message));
  }
  if (errors.length > 0) {
    console.warn({ message: 'GraphqlGateway request error', method, errors: errors.map((e) => e.message) });
    throw new GraphqlGatewayError(method, errors);
  }
};

export const createComponent = async (cloudId: string, project: ImportableProject): Promise<Component | never> => {
  const { name, description, type, labels, url } = project;
  const formattedLabels = labels.map((label) => label.split(' ').join('-').toLowerCase());
  const component = {
    name,
    description,
    type,
    labels: [IMPORT_LABEL, ...formattedLabels],
    links: [
      {
        type: CompassLinkType.Repository,
        url,
      },
    ],
    externalAlias: {
      externalId: project.id.toString(),
      externalSource: EXTERNAL_SOURCE,
    },
    cloudId,
  };

  const { data, errors } = await graphqlGateway.compass.asApp().createComponent(component);

  throwIfErrors('createComponent', errors);

  return data.component;
};

export async function getComponent(componentId: string): Promise<Component | never> {
  const { data, errors } = await graphqlGateway.compass
    .asApp()
    .getComponent({ componentId, options: { includeLinks: true } });

  throwIfErrors('getComponent', errors);
  return data.component;
}

export async function updateComponent(input: UpdateComponentInput): Promise<Component | never> {
  const { data, errors } = await graphqlGateway.compass.asApp().updateComponent(input);

  throwIfErrors('updateComponent', errors);
  return data.component;
}

export async function createExternalAlias(input: CreateCompassComponentExternalAliasInput): Promise<void | never> {
  const { errors } = await graphqlGateway.compass.asApp().createExternalAlias(input);
  throwIfErrors('createExternalAlias', errors);
}

export async function updateDataManager(input: UpdateCompassComponentDataManagerMetadataInput): Promise<void | never> {
  const { errors } = await graphqlGateway.compass.asApp().updateDataManager(input);
  throwIfErrors('updateDataManager', errors);
}

export async function detachDataManager(input: DetachCompassComponentDataManagerInput): Promise<void | never> {
  const { errors } = await graphqlGateway.compass.asApp().detachDataManager(input);
  throwIfErrors('detachDataManager', errors);
}

export async function deleteExternalAlias(input: DeleteCompassComponentExternalAliasInput): Promise<void | never> {
  const { errors } = await graphqlGateway.compass.asApp().deleteExternalAlias(input);

  if (errors.length === 1 && errors[0].message === UNKNOWN_EXTERNAL_ALIAS_ERROR_MESSAGE) {
    console.log('Could not find external alias to delete.');
    return;
  }

  throwIfErrors('deleteExternalAlias', errors);
}

export async function unlinkCompassComponents(cloudId: string, ecosystemAppId: string): Promise<void | never> {
  const { errors } = await graphqlGateway.compass.asApp().unlinkExternalSource({
    cloudId,
    ecosystemAppId,
    externalSource: EXTERNAL_SOURCE,
  });

  throwIfErrors('unlinkExternalSource', errors);
}

export async function getComponentByExternalAlias(input: GetComponentByExternalAliasInput): Promise<ComponentPayload> {
  const { errors, data } = await graphqlGateway.compass
    .asApp()
    .getComponentByExternalAlias({ ...input, externalSource: EXTERNAL_SOURCE });

  if (errors[0]?.message === COMPASS_GATEWAY_MESSAGES.COMPONENT_NOT_FOUND) {
    return { component: null };
  }

  throwIfErrors('getComponentByExternalAlias', errors);

  return data;
}

export async function sendEvents(eventPayload: CompassCreateEventInput | CompassCreateEventInput[]): Promise<void> {
  const { errors, data } = await graphqlGateway.compass.asApp().createEvent(eventPayload);

  throwIfErrors('createEvent', errors);

  return data;
}

export async function insertMetricValueByExternalId(cloudId: string, projectID: string, metric: Metric): Promise<void> {
  const { errors, data } = await graphqlGateway.compass.asApp().insertMetricValueByExternalId({
    cloudId,
    externalMetricSourceId: projectID,
    metricDefinitionId: metric.metricAri,
    value: {
      value: metric.value,
      timestamp: metric.timestamp ?? new Date().toISOString(),
    },
  });

  throwIfErrors('insertMetricValueByExternalId', errors);

  return data;
}
