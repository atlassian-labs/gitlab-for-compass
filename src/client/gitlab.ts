import { fetch } from '@forge/api';
import yaml from 'js-yaml';

import { BASE_URL } from '../constants';
import {
  GitlabAPIGroup,
  GroupAccessToken,
  RegisterWebhookPayload,
  CommitFileDiff,
  CompassYaml,
  GitlabAPIProject,
  ProjectBranch,
  MergeRequestState,
  MergeRequest,
  GitlabApiPipeline,
  Deployment,
  Environment,
  GitlabPipelineStates,
} from '../types';
import { GitlabHttpMethodError, InvalidConfigFileError } from '../models/errors';
import { INVALID_YAML_ERROR } from '../models/error-messages';
import { queryParamsGenerator } from '../utils/url-utils';

export enum HttpMethod {
  GET = 'GET',
  POST = 'POST',
  PUT = 'PUT',
  PATCH = 'PATCH',
  DELETE = 'DELETE',
}

export enum GitLabContentType {
  JSON = 'application/json',
  RAW = 'text/plain; charset=utf-8',
}
type CallGitLabConfig = {
  method?: HttpMethod;
  contentType?: GitLabContentType;
};

export enum GitLabHeaders {
  PAGINATION_TOTAL = 'x-total',
}

export type GitlabPaginatedFetch<D, P> = (
  page: number,
  perPage: number,
  fetchFnParameters: Record<'groupToken', string> & P,
) => Promise<{ data: D[]; headers: Headers }>;

export enum MergeRequestWorkInProgressFilterOptions {
  ONLY_WIP = 'yes',
  FILTER_OUT_WIP = 'no',
}

export enum MergeRequestResultViewOptions {
  SIMPLE = 'simple',
  DEFAULT = 'default',
}

export const callGitlab = async (
  path: string,
  authToken: string,
  config?: CallGitLabConfig,
  body?: string,
): Promise<any> => {
  const resp = await fetch(`${BASE_URL}${path}`, {
    method: config?.method || HttpMethod.GET,
    headers: {
      'PRIVATE-TOKEN': authToken,
      Accept: config?.contentType || GitLabContentType.JSON,
    },
    body,
  });

  if (resp.status === 204) {
    // no content, we can just return here
    return null;
  }

  if (resp.status >= 300) {
    console.warn(`Gitlab client received a status code of ${resp.status} while fetching ${path}`);
    throw new GitlabHttpMethodError(resp.status, resp.statusText);
  }

  if (config?.contentType === GitLabContentType.RAW) {
    return resp.text();
  }

  return { data: await resp.json(), headers: resp.headers };
};

export const getGroupsData = async (
  groupAccessToken: string,
  owned?: string,
  minAccessLevel?: number,
): Promise<GitlabAPIGroup[]> => {
  const params = {
    ...(owned ? { owned } : {}),
    ...(minAccessLevel ? { min_access_level: minAccessLevel.toString() } : {}),
  };

  const queryParams = queryParamsGenerator(params);

  const { data } = await callGitlab(`/api/v4/groups?${queryParams}`, groupAccessToken);

  return data;
};

export const registerGroupWebhook = async (payload: RegisterWebhookPayload): Promise<number> => {
  const { groupId, token: groupToken, url, signature } = payload;
  const {
    data: { id },
  } = await callGitlab(
    `/api/v4/groups/${groupId}/hooks`,
    groupToken,
    { method: HttpMethod.POST },
    JSON.stringify({
      url,
      token: signature,
      push_events: true,
      merge_requests_events: true,
      pipeline_events: true,
      deployment_events: true,
    }),
  );

  return id;
};

export const deleteGroupWebhook = async (groupId: number, hookId: number, groupToken: string): Promise<void> => {
  try {
    await callGitlab(`/api/v4/groups/${groupId}/hooks/${hookId}`, groupToken, { method: HttpMethod.DELETE });
  } catch (e) {
    if (e.message.includes('Not Found')) {
      return;
    }
    throw e;
  }
};

export const getGroupWebhook = async (
  groupId: number,
  hookId: number,
  groupToken: string,
): Promise<{ id: number } | null> => {
  try {
    const { data: webhook } = await callGitlab(`/api/v4/groups/${groupId}/hooks/${hookId}`, groupToken);

    return webhook;
  } catch (e) {
    if (e.message.includes('Not Found')) {
      return null;
    }
    throw e;
  }
};

export const getGroupAccessTokens = async (groupToken: string, groupId: number): Promise<GroupAccessToken[]> => {
  const { data: groupAccessTokenList } = await callGitlab(`/api/v4/groups/${groupId}/access_tokens`, groupToken);

  return groupAccessTokenList;
};

export const getCommitDiff = async (groupToken: string, projectId: number, sha: string): Promise<CommitFileDiff[]> => {
  const { data: diff } = await callGitlab(`/api/v4/projects/${projectId}/repository/commits/${sha}/diff`, groupToken);

  return diff;
};

export const getFileContent = async (
  groupToken: string,
  projectId: number,
  filePath: string,
  ref: string,
): Promise<CompassYaml> => {
  const params = {
    ref,
  };

  const queryParams = queryParamsGenerator(params);

  const fileRaw = await callGitlab(
    `/api/v4/projects/${projectId}/repository/files/${encodeURIComponent(filePath)}/raw?${queryParams}`,
    groupToken,
    { contentType: GitLabContentType.RAW },
  );

  try {
    return yaml.load(fileRaw);
  } catch (e) {
    console.warn({ message: 'Error parsing yaml file', error: e });
    throw new InvalidConfigFileError([INVALID_YAML_ERROR]);
  }
};

export const getProjects = async (
  groupToken: string,
  groupId: number,
  page: number,
  perPage: number,
  search?: string,
): Promise<{ data: GitlabAPIProject[]; headers: Headers }> => {
  const params = {
    include_subgroups: 'true',
    page: page.toString(),
    per_page: perPage.toString(),
    ...(search ? { search } : {}),
  };

  const queryParams = queryParamsGenerator(params);
  const { data, headers } = await callGitlab(`/api/v4/groups/${groupId}/projects?${queryParams}`, groupToken);

  return { data, headers };
};

export const getProjectById = async (groupToken: string, projectId: number): Promise<GitlabAPIProject> => {
  const { data: project } = await callGitlab(`/api/v4/projects/${projectId}`, groupToken);

  return project;
};

export const searchFileByPath = async (groupToken: string, projectId: number, path: string, branch: string) => {
  const params = {
    ref: branch,
  };

  const queryParams = queryParamsGenerator(params);

  const file = await callGitlab(`/api/v4/projects/${projectId}/repository/files/${path}?${queryParams}`, groupToken);

  return file;
};

export const getProjectLanguages = async (groupToken: string, projectId: number) => {
  const { data: languages } = await callGitlab(`/api/v4/projects/${projectId}/languages`, groupToken);

  return languages;
};

export const getProjectVariable = async (
  groupToken: string,
  projectId: number,
  variable: string,
): Promise<string | null | never> => {
  const {
    data: { value },
  } = await callGitlab(`/api/v4/projects/${projectId}/variables/${variable}`, groupToken);
  return value;
};

export const getProjectBranch = async (
  groupToken: string,
  projectId: number,
  branchName: string,
): Promise<ProjectBranch> => {
  const { data: branch } = await callGitlab(
    `/api/v4/projects/${projectId}/repository/branches/${branchName}`,
    groupToken,
  );
  return branch;
};

export const getOwnedProjectsBySearchCriteria = async (
  search: string,
  groupToken: string,
): Promise<GitlabAPIProject[]> => {
  const params = {
    owned: 'true',
    search,
  };

  const queryParams = queryParamsGenerator(params);

  const { data } = await callGitlab(`/api/v4/projects?${queryParams}`, groupToken);

  return data;
};

export const getProjectRecentDeployments: GitlabPaginatedFetch<
  Deployment,
  {
    projectId: number;
    dateAfter: string; // in iso date time format, e.g. 2019-03-15T08:00:00Z
    environmentName: string;
    dateBefore?: string; // in iso date time format, e.g. 2019-03-15T08:00:00Z
  }
> = async (page, perPage, fetchParameters) => {
  const { groupToken, projectId, dateAfter, dateBefore, environmentName } = fetchParameters;
  const params = {
    updated_after: dateAfter,
    environment: environmentName,
    page: page.toString(),
    per_page: perPage.toString(),
    ...(dateBefore ? { updated_before: dateBefore } : {}),
  };

  const queryParams = queryParamsGenerator(params);
  const path = `/api/v4/projects/${projectId}/deployments?${queryParams}`;

  const { data, headers } = await callGitlab(path, groupToken);

  return { data, headers };
};

export const getMergeRequests: GitlabPaginatedFetch<
  MergeRequest,
  {
    groupToken: string;
    projectId: number;
    state: MergeRequestState;
    scope: string;
    targetBranch?: string;
    orderBy?: string;
    wip?: MergeRequestWorkInProgressFilterOptions;
    isSimpleView?: boolean;
    sourceBranch?: string;
  }
> = async (page, perPage, fetchParameters) => {
  const { state, scope, targetBranch, orderBy, wip, isSimpleView, projectId, groupToken, sourceBranch } =
    fetchParameters;

  const params = {
    state,
    page: page.toString(),
    per_page: perPage.toString(),
    scope,
    ...(targetBranch ? { target_branch: targetBranch } : {}),
    ...(orderBy ? { order_by: orderBy } : {}),
    ...(wip ? { wip } : {}),
    ...(isSimpleView ? { view: 'simple' } : {}),
    ...(sourceBranch ? { source_branch: sourceBranch } : {}),
  };

  const queryParams = queryParamsGenerator(params);
  const path = `/api/v4/projects/${projectId}/merge_requests?${queryParams}`;

  const { data, headers } = await callGitlab(path, groupToken);

  return { data, headers };
};

export const getProjectDeploymentById = async (projectId: number, deploymentId: number, groupToken: string) => {
  const { data } = await callGitlab(`/api/v4/projects/${projectId}/deployments/${deploymentId}`, groupToken);

  return data;
};

export const getEnvironments = async (projectId: number, groupToken: string): Promise<Environment[]> => {
  const { data } = await callGitlab(`/api/v4/projects/${projectId}/environments`, groupToken);

  return data;
};

export const getProjectRecentPipelines: GitlabPaginatedFetch<
  GitlabApiPipeline,
  {
    projectId: number;
    dateAfter?: string; // in iso date time format, e.g. 2019-03-15T08:00:00Z
    branchName: string;
    status?: GitlabPipelineStates;
  }
> = async (page, perPage, fetchParameters) => {
  const { groupToken, projectId, dateAfter, branchName, status } = fetchParameters;
  const params = {
    ref: branchName,
    page: page.toString(),
    per_page: perPage.toString(),
    ...(status ? { status } : {}),
    ...(dateAfter ? { updated_after: dateAfter } : {}),
  };

  const queryParams = queryParamsGenerator(params);
  const path = `/api/v4/projects/${projectId}/pipelines?${queryParams}`;

  const { data, headers } = await callGitlab(path, groupToken);

  return { data, headers };
};

export const createFileInProject = async (
  groupToken: string,
  projectId: number,
  filePath: string,
  branchName: string,
  startBranchName: string,
  encoding: string,
  content: string,
  commitMessage: string,
) => {
  const path = `/api/v4/projects/${projectId}/repository/files/${filePath}`;

  const { data } = await callGitlab(
    path,
    groupToken,
    { method: HttpMethod.POST },
    JSON.stringify({
      branch: branchName,
      start_branch: startBranchName,
      encoding,
      content,
      commit_message: commitMessage,
    }),
  );

  return data;
};

export const createMergeRequest = async (
  groupToken: string,
  projectId: number,
  sourceBranch: string,
  targetBranch: string,
  title: string,
  description: string,
  removeSourceBranch: boolean,
) => {
  const path = `/api/v4/projects/${projectId}/merge_requests`;

  const { data } = await callGitlab(
    path,
    groupToken,
    { method: HttpMethod.POST },
    JSON.stringify({
      source_branch: sourceBranch,
      target_branch: targetBranch,
      title,
      description,
      remove_source_branch: removeSourceBranch,
    }),
  );

  return data;
};
