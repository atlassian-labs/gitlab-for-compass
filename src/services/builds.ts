import { CompassBuildEventState, CompassCreateEventInput, DataProviderBuildEvent } from '@atlassian/forge-graphql';
import { max } from 'lodash';

import { GitlabPipelineStates, GitlabApiPipeline, PipelineEvent } from '../types';

export const pipelineLastUpdated = (pipeline: PipelineEvent): Date => {
  if (pipeline.object_attributes.finished_at !== null) {
    return new Date(pipeline.object_attributes.finished_at);
  }
  let latestTime = new Date(pipeline.object_attributes.created_at).getTime();
  for (const build of pipeline.builds) {
    latestTime = max([new Date(build.started_at).getTime(), new Date(build.finished_at).getTime(), latestTime]);
  }
  return new Date(latestTime);
};

export const toCompassBuildState = (state: string): CompassBuildEventState => {
  switch (state) {
    case GitlabPipelineStates.CREATED:
    case GitlabPipelineStates.WAITING_FOR_RESOURCE:
    case GitlabPipelineStates.PREPARING:
    case GitlabPipelineStates.PENDING:
    case GitlabPipelineStates.RUNNING:
    case GitlabPipelineStates.MANUAL:
      return CompassBuildEventState.InProgress;
    case GitlabPipelineStates.SUCCESS:
      return CompassBuildEventState.Successful;
    case GitlabPipelineStates.FAILED:
      return CompassBuildEventState.Failed;
    case GitlabPipelineStates.CANCELED:
      return CompassBuildEventState.Cancelled;
    case GitlabPipelineStates.SCHEDULED:
      return CompassBuildEventState.Unknown;
    default:
      return CompassBuildEventState.Unknown;
  }
};

export const webhookPipelineEventToCompassBuildEvent = (
  pipeline: PipelineEvent,
  cloudId: string,
): CompassCreateEventInput => {
  const lastUpdated = pipelineLastUpdated(pipeline);
  return {
    cloudId,
    event: {
      build: {
        externalEventSourceId: pipeline.project.id.toString(),
        updateSequenceNumber: lastUpdated.getTime(),
        displayName: `${pipeline.project.name} pipeline ${pipeline.object_attributes.id}`,
        description: `Pipeline run ${pipeline.object_attributes.id} for project ${pipeline.project.name}`,
        url: `${pipeline.project.web_url}/-/pipelines/${pipeline.object_attributes.id}`,
        lastUpdated: lastUpdated.toISOString(),
        buildProperties: {
          pipeline: {
            pipelineId: pipeline.object_attributes.id.toString(),
          },
          state: toCompassBuildState(pipeline.object_attributes.status),
          startedAt: new Date(pipeline.object_attributes.created_at).toISOString(),
          completedAt: new Date(pipeline.object_attributes.finished_at).toISOString(),
        },
      },
    },
  };
};

export const gitlabApiPipelineToCompassDataProviderBuildEvent = (
  pipeline: GitlabApiPipeline,
  projectName: string,
): DataProviderBuildEvent => {
  const isCompleted = !(toCompassBuildState(pipeline.status) === CompassBuildEventState.InProgress);

  return {
    description: `Pipeline run ${pipeline.id} for project ${projectName}`,
    displayName: `${projectName} pipeline ${pipeline.id}`,
    state: toCompassBuildState(pipeline.status),
    startedAt: new Date(pipeline.created_at).toISOString(),
    completedAt: isCompleted ? new Date(pipeline.updated_at).toISOString() : null,
    lastUpdated: new Date(pipeline.updated_at).toISOString(),
    updateSequenceNumber: new Date(pipeline.updated_at).getTime(),
    url: pipeline.web_url,
    pipeline: {
      pipelineId: pipeline.id.toString(),
    },
  };
};
