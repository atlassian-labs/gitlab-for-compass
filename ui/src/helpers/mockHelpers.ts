import { invoke as realInvoke, view } from '@forge/bridge';

export const invoke: jest.Mock = realInvoke as jest.Mock;
const getContext: jest.Mock = view.getContext as jest.Mock;

export const defaultMocks: {
  [key: string]: unknown;
} = {
  groups: {
    success: true,
    data: [],
  },
  features: {
    success: true,
    data: {},
  },
  appId: {
    success: true,
    data: 'app-id',
  },
};

export const mockInvoke = (mocks = defaultMocks) => {
  invoke.mockImplementation(async (key) => {
    if (mocks[key] instanceof Error) {
      throw mocks[key];
    }

    return mocks[key];
  });
};

export const mockGetContext = (moduleKey: string) => {
  getContext.mockImplementation(async () => {
    return {
      moduleKey,
    };
  });
};
