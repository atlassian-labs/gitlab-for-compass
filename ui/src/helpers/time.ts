export const formatLastSyncTime = (lastSyncTime: string): string => {
  return `${new Date(lastSyncTime).toLocaleDateString('en-US')} ${new Date(lastSyncTime).toLocaleTimeString('en-US')}`;
};
