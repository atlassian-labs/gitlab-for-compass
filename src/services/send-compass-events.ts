import { CompassCreateEventInput } from '@atlassian/forge-graphql';
import { sendEvents } from '../client/compass';

export const sendEventToCompass = async (payload: CompassCreateEventInput): Promise<void> => {
  try {
    await sendEvents(payload);
  } catch (e) {
    console.error(`Error sending event to Compass`, e);
  }
};
