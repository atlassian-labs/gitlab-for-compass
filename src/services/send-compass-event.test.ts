/* eslint-disable import/first */
/* eslint-disable import/order */
import { mocked } from 'jest-mock';
import { mockForgeApi } from '../__tests__/helpers/forge-helper';

mockForgeApi();

import { CompassEventType } from '@atlassian/forge-graphql';
import { sendEvents } from '../client/compass';
import { sendEventToCompass } from './send-compass-events';
import { pipelineWebhookFixture } from '../__tests__/fixtures/build-webhook-payload';
import { MOCK_CLOUD_ID } from '../__tests__/fixtures/gitlab-data';
import { webhookPipelineEventToCompassBuildEvent } from './builds';

jest.mock('../client/compass');
const mockSendEvents = mocked(sendEvents);

describe('send build event to compass method', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('successfully maps gitlab pipeline event to compass build event and send it', async () => {
    await sendEventToCompass(webhookPipelineEventToCompassBuildEvent(pipelineWebhookFixture, MOCK_CLOUD_ID));

    expect(mockSendEvents).toMatchSnapshot();
  });
});
