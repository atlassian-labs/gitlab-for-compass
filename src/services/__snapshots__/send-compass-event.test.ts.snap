// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`send build event to compass method successfully maps gitlab pipeline event to compass build event and send it 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "cloudId": "0a44684d-52c3-4c0c-99f8-9d89ec294759",
        "event": Object {
          "build": Object {
            "buildProperties": Object {
              "completedAt": "2022-06-14T22:46:49.000Z",
              "pipeline": Object {
                "pipelineId": "563913989",
              },
              "startedAt": "2022-06-14T22:45:56.000Z",
              "state": "SUCCESSFUL",
            },
            "description": "Pipeline run 563913989 for project stringray",
            "displayName": "stringray pipeline 563913989",
            "externalEventSourceId": "36867443",
            "lastUpdated": "2022-06-14T22:46:49.000Z",
            "updateSequenceNumber": 1655246809000,
            "url": "https://gitlab.com/gitlab-com/alliances/atlassian/shared-projects/compass-electromagnets-testing/patrick-subgroup/stringray/-/pipelines/563913989",
          },
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
